[Feature][Integration][Java] add ElasticsearchVectorStore in Java

twosom opened the pull request:
<!--
* Thank you very much for contributing to Flink Agents.
* Please add the relevant components in the PR title. E.g., [api], [runtime], [java], [python], [hotfix], etc.
-->

<!-- Please link the PR to the relevant issue(s). Hotfix doesn't need this. -->
Linked issue: #321

### Purpose of change

This PR implements the Java version of Vector Store functionality for Flink Agents, following the design proposal in #143. This implementation enables RAG (Retrieval-Augmented Generation) capabilities by providing vector-based context retrieval.

- Added Vector Store API with query support and document handling
- Added `@VectorStore` annotation for agent plan resource management
- Added context retrieval request and response events
- Implemented context retrieval action with vector store support
- Implemented Elasticsearch vector store integration
- Added ElasticsearchVectorStore RAG example with Ollama integration

### Tests

<!-- How is this change verified? -->

### API

<!-- Does this change touches any public APIs? -->

### Documentation

<!-- Do not remove this section. Check the proper box only. -->

- [x] `doc-needed` <!-- Your PR changes impact docs -->
- [ ] `doc-not-needed` <!-- Your PR changes do not impact docs -->

wenjin272 commented:
Hi, @twosom, I'll start reviewing this PR as soon as possible. Besides, this PR should be labeled 'doc-needed'.

wenjin272 commented on plan/src/main/java/org/apache/flink/agents/plan/actions/ContextRetrievalAction.java:
```diff
@@ -0,0 +1,75 @@
+                        new Class[] {Event.class, RunnerContext.class}),
+                List.of(
+                        ContextRetrievalRequestEvent.class.getName(),
+                        ContextRetrievalResponseEvent.class.getName()));
```
This action expect receive `ContextRetrievalRequestEvent` and send `ContextRetrievalResponseEvent `, so the listen event types here should only be `ContextRetrievalRequestEvent.class.getName()`

wenjin272 commented on integrations/vector-stores/elasticsearch/src/main/java/org/apache/flink/agents/integrations/vectorstores/elasticsearch/ElasticsearchVectorStore.java:
```diff
@@ -0,0 +1,308 @@
+        final RestClient restClient = builder.build();
+        final ElasticsearchTransport transport =
+                new RestClientTransport(restClient, new JacksonJsonpMapper());
+        this.client = new ElasticsearchClient(transport);
```
Here create a client but no where close it. This may cause connection leak. But this is a common issue for resources using persistent connection, and need framework provide clean up mechanism. I will create a dedicated issue for this.

wenjin272 reviewed (commented):
Hi, @twosom, thanks for your contribution. Overall looks good to me. I left some minor comments about details.

Besides, due to the example requires es cluster, I will verify the example next week.

And some test cases failed, I think it is caused by the newly added built-in action, which break the agent plan verification.

twosom commented on integrations/vector-stores/elasticsearch/src/main/java/org/apache/flink/agents/integrations/vectorstores/elasticsearch/ElasticsearchVectorStore.java:
```diff
@@ -0,0 +1,308 @@
+            String filter = (String) args.get("filter_query");
+
+            List<Float> queryVector = new ArrayList<>(embedding.length);
+            for (float v : embedding) queryVector.add(v);
```
@wenjin272 
Thank you for your feedback.
But there is no float support in `Arrays.stream()` method.

<img width="515" height="285" alt="image" src="https://github.com/user-attachments/assets/7f24588b-0abd-4c66-9d49-866ffce17830" />

wenjin272 replied:
You are right, sorry for this suggestion.

wenjin272 commented on api/src/main/java/org/apache/flink/agents/api/vectorstores/BaseVectorStore.java:
```diff
@@ -0,0 +1,96 @@
+
+        // TODO
+        // for now, we don't need to use additional parameters.
+        final float[] queryEmbedding = embeddingModel.embed(query.getQueryText(), Map.of());
```
Could directly use `embeddingModel.embed(query.getQueryText())` here.

wenjin272 commented on integrations/vector-stores/elasticsearch/src/main/java/org/apache/flink/agents/integrations/vectorstores/elasticsearch/ElasticsearchVectorStore.java:
```diff
@@ -0,0 +1,308 @@
+            }
+
+            final SearchResponse<Map<String, Object>> searchResponse =
+                    (SearchResponse) this.client.search(builder.build(), Map.class);
```
I think this make sense to some extent. But if different implementation return document in different structure, the user application must be aware of the implementation they used. If they changed the vector store used in their  code, they may also change the code which process the returned documents.
WDTY @alnzng ?

wenjin272 replied:
I think this make sense to some extent. But if different implementation return document in different structure, the user application must be aware of the implementation they used. If they changed the vector store used in their  code, they may also change the code which process the returned documents.
WDTY @alnzng, @xintongsong ?

wenjin272 reviewed (commented):
Hi, @twosom, thanks for your update. I left some comments mainly about the document type.

Additionally, when run the example locally, I encountered an issue related to SSL certificate authentication. I didn't seem to find where we can configure the `http_ca.crt`, so I manually disabled security authentication for the es cluster.

wenjin272 commented:
> @wenjin272 Thank you for your review.
> 
> I didn't consider SSL configuration while testing locally. Thank you for pointing this out.
> 
> Would it be better to add SSL certificate configuration support in a separate PR rather than including it in this one? I'm happy to follow your guidance on this.
> 
> Thank you.

I think ssl support can be done in separate PR.

xintongsong commented on examples/src/main/java/org/apache/flink/agents/examples/rag/ElasticsearchRagExample.java:
```diff
@@ -0,0 +1,431 @@
+ *   <li>Print the model's answers for a set of example queries.
+ * </ol>
+ */
+public class ElasticsearchRagExample {
```
And it's not necessary to include 2) in this PR. We can make it a separate follow-up task.

twosom commented on examples/src/main/java/org/apache/flink/agents/examples/rag/ElasticsearchRagExample.java:
```diff
@@ -0,0 +1,431 @@
+                Object content = docs.get(i).getContent();
+                items.add(String.format("%d. %s", i + 1, content));
+            }
+            String contextText = String.join("\n\n", items);
```
@xintongsong

Thank you for your feedback.
I completely agree with your point.

If we use strings, I think there are two approaches:

Serialize the entire Map to a string
Specify which field(s) to use for LLM context (e.g., title or content)
While option 1 (serializing the entire Map) isn't necessarily bad, unless we explicitly specify Elasticsearch's _source fields, all content would be used as LLM context, which could lead to significant token waste.

With option 2, we can provide only the necessary property (or properties) to the LLM, which would result in more appropriate token usage. However, this would add more initialization parameters to ElasticsearchVectorStore, increasing complexity.

I find it difficult to make this decision alone.

WDYT?
@xintongsong @wenjin272

xintongsong replied:
I think retrieving the entire map from ES does not necessarily mean feeding everything to LLM. We can apply transformations on the retrieved data before sending them to the LLM. This can be done in several ways:

- The user can add a custom Action that listens to the context retrieval response.
- We can introduce a built-in RagAgent that accepts a function which will be applied on the retrieved documents
- We can add a function parameter in the general vector store interface which will be applied on the retrieved data

With either of the above options, we don't need to do this specifically for each vector store implementation.

twosom commented:
@xintongsong @wenjin272 

Hello.
As you mentioned, I removed the generic type and changed it to a string type.
I also removed the existing examples and added e2e tests.
I would appreciate it if you could review it when you have time.
Thank you.

xintongsong commented on integrations/vector-stores/elasticsearch/pom.xml:
```diff
@@ -0,0 +1,52 @@
+            <groupId>org.apache.flink</groupId>
+            <artifactId>flink-agents-plan</artifactId>
+            <version>${project.version}</version>
+        </dependency>
```
I think we should not need to depends on `flink-agents-plan`?

wenjin272 commented on e2e-test/flink-agents-end-to-end-tests-integration/src/test/java/org/apache/flink/agents/integration/test/VectorStoreIntegrationTest.java:
```diff
@@ -0,0 +1,122 @@
+
+    @SuppressWarnings("unchecked")
+    private void checkResult(CloseableIterator<Object> results) {
+        // 단일 입력에 대한 결과 1건 검증 및 벤더 비의존 최소 계약 확인
```
comments not in English.
